### PR TITLE
[DataGrid] Support Checkbox component slot

### LINKED ISSUE #281: [DataGrid] Extend components API to support custom checkbox
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

Provide a new entry point in the `components` prop to customize the checkbox

## Examples 🌈

```jsx
<DataGrid
  components={{
    Checkbox: MyCheckbox
  }}
/>
```

Capture d’écran 2020-09-13 à 19 27 44

## Motivation 🔦

I was working on a customization example to match the Ant Design specification. Such an API could make customization easier. I'm not sure it's a frequent request, hence the "waiting for upvotes" label so we don't rush into implementation if it doesn't solve a strong pain.

### LINKED ISSUE #1438: Latest recommendation for customizing the DataGrid Checkbox
About six months ago the status for issue #281 was
>Snoozing until we get users reporting it :).

Since then, progress has been made on how to customize various components in the DataGrid.  However, I did not see anything for how to do so for the Checkbox.

What is the recommended approach to directly customize the checkbox? (i.e., without creating a separate column)

Thank you in advance!

\- E

### PR DESCRIPTION
**Pull request by ZeeshanTamboli**
Closes #281

### CODE REVIEW

**Comment by ZeeshanTamboli** [comment]
Would we want to create 2 separate component slots, one for checkbox header and one for checkbox cell? Or one?

Currently it is `GridCellCheckboxRenderer` for cells and `GridHeaderCheckbox` for checkbox header

**Comment by oliviertassinari** [comment]
@ZeeshanTamboli The initial problem I wanted to solve is the customization of the look and feel, not the functional behavior. It was about preparing the unstyled components, so the former option (1 component).

**Comment by ZeeshanTamboli** [comment]
Request for review so that if looks fine I can move on to cover it up with tests. Would it be worth it to have coverage for this?

**Review by oliviertassinari — COMMENTED** [review]
- It looks good at a high level
- Changes might conflict with #1552
- I had initially opened the issue for https://material-ui.com/components/data-grid/style, but not that I think about it, it feels like this customization should be implemented with the theme overrdes.

**Comment by ZeeshanTamboli** [comment]
> * I had initially opened the issue for https://material-ui.com/components/data-grid/style, but not that I think about it, it feels like this customization should be implemented with the theme overrdes.

Well I thought this was good to take as I saw some upvotes for this at https://github.com/mui-org/material-ui-x/issues/281#issue-700611756

**Comment by oliviertassinari** [comment]
@ZeeshanTamboli It's still bringing value, my point is that we likely don't want to use the new option in the theming demos.

**Comment by EdmundsEcho** [comment]
>... but not that I think about it, it feels like this customization should be implemented with the theme overrides.

Quick question on how this approach might work.
In order to customize the icon used for a checkbox the user can use the prop settings for the `Mui-Checkbox`. e.g., taken from the checkbox demo page:

```jsx
<Checkbox icon={<FavoriteBorder />} checkedIcon={<Favorite />} name="checkedH" />
```
1. Is there a way to control these entries using the style overrides prop?
2. If a user wanted to customize the prop in our theme, is there a mechanism for selecting/targeting the checkbox for `Mui-DataGrid`? (i.e., avoiding just using a default that impact the app globally)

**Comment by oliviertassinari** [comment]
@EdmundsEcho if you only want to change the look and feel of the checkboxes in the data grid, and not the whole app, using a components prop once this PR land is probably the most effective. Otherwise, you could nest the theme providers.

**Comment by EdmundsEcho** [comment]
@oliviertassinari  Thank you for the response.  

Has it been confirmed/is there any reason to believe the data grid does not read the default props of the custom theme?  I ask because the following entry is not being applied to the data-grid:
```js
...
   props: {
     
      MuiButtonBase: {
        disableRipple: true, 
      },
      
      MuiCheckbox: {
        disableRipple: true,
        checkedIcon: <Check />,
        icon: <Clear />,
      },
    ...
   }
}
```

**Comment by ZeeshanTamboli** [comment]
@EdmundsEcho If you want to check with `components` prop for your custom `Checkbox` _I think_ you can check with this npm package https://pkg.csb.dev/mui-org/material-ui-x/commit/2e9da7f6/@material-ui/data-grid which includes this PR code.

Also I think `MuiCheckbox` at theme override works because the default checkbox is MUI's checkbox for data grid checkbox.

cc: @oliviertassinari

File: packages/grid/_modules_/grid/components/columnSelection/GridCellCheckboxRenderer.tsx [hunk 1]
```diff
@@ -7,48 +6,56 @@ import { GridCellParams } from '../../models/params/gridCellParams';
 import { isNavigationKey, isSpaceKey } from '../../utils/keyboardUtils';
 import { GridApiContext } from '../GridApiContext';
 
-export const GridCellCheckboxRenderer = React.memo((props: GridCellParams) => {
-  const { getValue, field, id } = props;
-  const apiRef = React.useContext(GridApiContext);
-  const tabIndexState = useGridSelector(apiRef, gridTabIndexCellSelector);
+const GridCellCheckboxForwardRef = React.forwardRef<HTMLInputElement, GridCellParams>(
+  function GridCellCheckbox(props, ref) {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnSelection/GridCellCheckboxRenderer.tsx:10]:
```suggestion
const GridCellCheckboxForwardRef = React.forwardRef<HTMLInputElement, GridCellParams>(
  function GridCellCheckboxRenderer(props, ref) {
```

**Review by oliviertassinari — DISMISSED** [review]

**Comment by EdmundsEcho** [comment]
@ZeeshanTamboli  Thank you for the FYI.  I greatly appreciate it.  I will give the version you pointed me to later this week.  In the meantime, I'm somewhat surprised to learn that I can use the `overrides` key to change the *components themselves* used for both checked and unchecked states.  Said differently, is there a way to use the `overrides` key to set the component for the unchecked state to `<Clear />` from '@material-ui/icons/Clear` for the `MuiCheckbox`?

**Comment by ZeeshanTamboli** [comment]
> @ZeeshanTamboli Thank you for the FYI. I greatly appreciate it. I will give the version you pointed me to later this week. In the meantime, I'm somewhat surprised to learn that I can use the `overrides` key to change the _components themselves_ used for both checked and unchecked states. Said differently, is there a way to use the `overrides` key to set the component for the unchecked state to `<Clear />` from '@material-ui/icons/Clear`for the`MuiCheckbox`?

If you want to customize the checkbox of data grid in any way you want it would be better to use the `components` prop and use `Checkbox` slot of data grid once this PR lands.
As for global theme overrides (all checkboxes in global app) since you tried and it works, you can check the (Checkbox API)(https://material-ui.com/api/checkbox/) for more details on each prop.

>is there a way to use the `overrides` key to set the component for the unchecked state to `<Clear />` 

I think for this you can use `icon` prop of checkbox. The icon to display when the component is unchecked.
Please create a separate issue for more questions/issues apart from this PR.

**Comment by EdmundsEcho** [comment]
Got it.  That is my understanding too.  I just wanted to make sure I did not miss something; there is no way to set the icon prop using the overrides key, but you can do so using the props key.  Thank you!!

**Comment by ZeeshanTamboli** [comment]
Just commenting to let you'll know this was ready for review.

**Review by dtassone — APPROVED** [review]

**Comment by dtassone** [comment]
Nice and clean 👍

**Comment by ZeeshanTamboli** [comment]
Thanks. If we are considering #1552, it needs to be merged before this, to make some type changes here as well.

**Comment by dtassone** [comment]
I don't think #1152 is a blocker, this one is ready so merging now. We can rebase #1152 and do the change there.

**Comment by oliviertassinari** [comment]
> I don't think #1152 is a blocker

@dtassone This old close issue doesn't seem to be related, why bring it up? Ah no ok, I think that it's related to #1552.

**Pull request merged**